### ISSUE TITLE
Sometimes Entity AI seems not working (not moving / not attackable)

### ISSUE THREAD

**Issue by kniffo80** [RevivalPMMP/PureEntitiesX#64]
I'm using **PureEntitiesX** with the **mcpe-1.0** branch of PocketMine-MP and the latest **1.0.0-bug-fixes** branch of PureEntitiesX.

After some time, nearly all entities don't move anymore for a period of time. They're not _attackable_ anymore. I didn't see any clue in the log causing this error. But maybe, this is a known issue.

Reproducable: regulary, but not clue about the error that "freezes" the mobs.

**Side-note:** one of the best Mob plugins found.

Post-Note: I have some CRITS in my server log:

`[15:57:33] [Server thread/CRITICAL]: InvalidArgumentException: "Value 18 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[15:57:33] [Server thread/CRITICAL]: Could not tick level "can": Value 16 exceeds the range!
[15:57:33] [Server thread/CRITICAL]: InvalidArgumentException: "Value 16 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[15:57:33] [Server thread/CRITICAL]: Could not tick level "can": Value 14 exceeds the range!
[15:57:33] [Server thread/CRITICAL]: InvalidArgumentException: "Value 14 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[15:57:33] [Server thread/CRITICAL]: Could not tick level "can": Value 12 exceeds the range!
[15:57:33] [Server thread/CRITICAL]: InvalidArgumentException: "Value 12 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[15:57:33] [Server thread/CRITICAL]: Could not tick level "can": Value 10 exceeds the range!
[15:57:33] [Server thread/CRITICAL]: InvalidArgumentException: "Value 10 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:09] [Server thread/CRITICAL]: Could not tick level "can": Value 10 exceeds the range!
[16:00:09] [Server thread/CRITICAL]: InvalidArgumentException: "Value 10 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 19 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 19 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 18 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 18 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 17 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 17 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 16 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 16 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 15 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 15 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 14 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 14 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 13 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 13 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 12 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 12 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
[16:00:24] [Server thread/CRITICAL]: Could not tick level "can": Value 11 exceeds the range!
[16:00:24] [Server thread/CRITICAL]: InvalidArgumentException: "Value 11 exceeds the range!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175
`
I know that this is related to PocketMine-MP - but maybe, those values have to be adjusted by the plugin?

**Comment by Sandertv** [comment]
Thanks for your report! I think this is because of a creeper trying to explode near you, but those still seem to lag a lot and give errors. I'll look into it!

**Comment by kniffo80** [comment]
Maybe ;)

I never saw a creeper explode. They only hit but don't explode - that's what i found out so far.

Maybe it's a good idea to have an own logfile for the plugin which writes debug stuff to the logfile? Good thing about the own logfile: it can be appended to a bug report or feature request ;)

**Comment by Sandertv** [comment]
Oh that error there, we are not sure what exactly causes this, but we're thinking it has todo with some memory things... Not too sure, we'll see what we can do.

**Comment by Sandertv** [comment]
And the logfile: I'm not exactly sure how we should do that, but it's quite a nice idea.

**Comment by kniffo80** [comment]
I'm not an experienced PHP programmer - only a JAVA programmer. I'll try to look into this issue for myself too.

**Comment by Sandertv** [comment]
How much RAM/cpu and stuff do you have?

**Comment by kniffo80** [comment]
I have a VPS with Ubuntu, 2 vCores and 1 GB of RAM. I have another server where nearly no services are running. I will try to deploy PocketMine-MP and the plugin to this server. Maybe, it's really a matter of free RAM. I don't know. Will report back when I've done so.

**Comment by kniffo80** [comment]
After moving whole stuff to my other server, which has 800MB memory left after starting PocketMine-MP - it's the same. I think it has nothing to do with RAM or CPU's ...

I don't like those bugs because they are hard to reproduce without a stack trace :(

**Comment by Sandertv** [comment]
In fact I think I just realized the problem!

**Comment by Sandertv** [comment]
It might not do anything, but could you try makig the spawning limit in the pocketmine.yml for the mobs higher?

**Comment by caspervanneck** [comment]
@kniffo80 do yo use genisys or forks

if so dont ese genisys forks

**Comment by kniffo80** [comment]
@Sandertv yes i can do so - which value should i try?
@caspervanneck I'm using "vanilla" PocketMine-MP (the 1.0 branch) no Genisys, no other forks. I've tried them all - PocketMine is the most clear

**Comment by caspervanneck** [comment]
hmm i dont know how to fix it

**Comment by kniffo80** [comment]
Regarding the logfile thingy. Add the following line to PureEntites.php class:

`    /**
     * Logs a logline to the plugin's logfile ...
     * @param string $logline	the logline to be appended
     */
    public static function logDebug (string $logline) {
    	file_put_contents('./pureentities_'.date("j.n.Y").'.log', "\033[32m".(date("j.n.Y G:i:s")." [DEBUG] ".$logline."\033[0m\r\n"), FILE_APPEND);
    }
    
    /**
     * Logs a logline to the plugin's logfile ...
     * @param string $logline	the logline to be appended
     */
    public static function logNormal (string $logline) {
    	file_put_contents('./pureentities_'.date("j.n.Y").'.log', "\033[37m".(date("j.n.Y G:i:s")." [INFO]  ".$logline."\033[0m\r\n"), FILE_APPEND);
    }
    
    
    /**
     * Logs a logline to the plugin's logfile ...
     * @param string $logline	the logline to be appended
     */
    public static function logWarn (string $logline) {
    	file_put_contents('./pureentities_'.date("j.n.Y").'.log', "\033[31m".(date("j.n.Y G:i:s")." [WARN]  ".$logline."\033[0m\r\n"), FILE_APPEND);
    }`

And then you can use the static methods simply with (e.g.):
`PureEntities::logWarn ("Testtest");`

I'm not invited to the group - so I cannot commit and push any changes right now ;)

**Comment by Sandertv** [comment]
Thanks for that!
Just try adding about 30 to all of them, especially animals are important I think.

**Comment by kniffo80** [comment]
No problem.

I've set max spawns of all to 100 - doesn't solve the problem :/

**Comment by Sandertv** [comment]
Alright. Guess that isn't the problem either then.

**Comment by kniffo80** [comment]
I'll try to debug that stuff tomorrow - for today I'm going to chill on my couch ;-)

**Comment by kniffo80** [comment]
I took a look at the Attribute class in PocketMine-MP, that throws these Exceptions. It seems that the min/max values defined in Attribute class are too high:

```
		self::addAttribute(self::ABSORPTION, "minecraft:absorption", 0.00, 340282346638528859811704183484516925440.00, 0.00);
		self::addAttribute(self::SATURATION, "minecraft:player.saturation", 0.00, 20.00, 5.00);
		self::addAttribute(self::EXHAUSTION, "minecraft:player.exhaustion", 0.00, 5.00, 0.41);
		self::addAttribute(self::KNOCKBACK_RESISTANCE, "minecraft:knockback_resistance", 0.00, 1.00, 0.00);
		self::addAttribute(self::HEALTH, "minecraft:health", 0.00, 20.00, 20.00);
		self::addAttribute(self::MOVEMENT_SPEED, "minecraft:movement", 0.00, 340282346638528859811704183484516925440.00, 0.10);
		self::addAttribute(self::FOLLOW_RANGE, "minecraft:follow_range", 0.00, 2048.00, 16.00, false);
		self::addAttribute(self::HUNGER, "minecraft:player.hunger", 0.00, 20.00, 20.00);
		self::addAttribute(self::ATTACK_DAMAGE, "minecraft:attack_damage", 0.00, 340282346638528859811704183484516925440.00, 1.00, false);
		self::addAttribute(self::EXPERIENCE_LEVEL, "minecraft:player.level", 0.00, 24791.00, 0.00);
		self::addAttribute(self::EXPERIENCE, "minecraft:player.experience", 0.00, 1.00, 0.00);
```

I don't know which attribute it is ... checking now ...

**Comment by kniffo80** [comment]
I found the following (after adding more debug to PocketMine-MP):

`[04:20:40] [Server thread/CRITICAL]: InvalidArgumentException: "Value 18 exceeds the range for minecraft:health!" (EXCEPTION) in "/src/pocketmine/entity/Attribute" at line 175`

It seems that the health attribute exceeds the max definition:

`self::addAttribute(self::HEALTH, "minecraft:health", 0.00, 20.00, 20.00);`

But as you can see, it's max value is 20.00 - now why does 18 exceed the range? Is it because of the definition is done in a float format and health itself is given as an integer?

**Comment by kniffo80** [comment]
Please update your entity creation. When setting max health, you need to set the health to max health. An example (for Wolf):

```
    public function initEntity(){
        parent::initEntity();

        if(isset($this->namedtag->Angry)){
            $this->angry = (int) $this->namedtag["Angry"];
        }

        $this->setMaxHealth(8);
        $this->setHealth(8);
        $this->fireProof = false;
        $this->setDamage([0, 3, 4, 6]);
    }
```

These exceptions don't occur anymore. But it has nothing to do with the "original" bug. Still investigating ...

**Comment by kniffo80** [comment]
LOL - I found out, that **LostPets** is not working quite good with PureEntitesX :smile: 

I deactivated the plugin and now everything seems to work quite fine.

But someone needs to adjust the entities the way i mentioned above. I will do a small checklist here what to be done:

- [ ] when setting maxHealth, also set health to prevent CRIT messages in server log
- [ ] when startTiming is done - sometimes the stopTiming is not called (instead, startTiming is called again)
- [ ] add plugin logfile (optional) for debugging reasons

I can start a pull request if wanted. I've the changes locally stored ;)
